src/ctap/mod.rs

```diff
@@ -2067,12 +2067,7 @@ mod test {
         env.customization_mut().enterprise_rp_id_list = vec!["example.com".to_string()];
         let mut ctap_state = CtapState::new(&mut env, CtapInstant::new(0));
 
-        let mut key_bytes = [0; 32];
-        let private_key = crypto::ecdsa::SecKey::gensk(env.rng());
-        private_key.to_bytes(array_mut_ref!(key_bytes, 0, 32));
-        storage::set_attestation_certificate(&mut env, &[0xCC]).unwrap();
-        storage::set_attestation_private_key(&mut env, &key_bytes).unwrap();
-        storage::enable_enterprise_attestation(&mut env).unwrap();
+        env.set_enterprise_attestation();
```

Comment:
I would argue that this should be done before calling `CtapState::new`. That would be a change with current code (so not related to this PR), but I think it makes more sense for tests with non-empty storage to not let `CtapState::new` initialize the storage. Since this cannot happen in practice that `CtapState` is created and the storage is modified under its feet.

At least something like that would make sense:

```rust
let _ = CtapState::new(&mut env, ...);  // initial boot of device
env.foo();  // simulate some modification without calling the actual functions
let mut ctap_state = CtapState::new(&mut env, ...);  // boot device for test
```

But ideally we want (in particular for fuzzing sequences):

```rust
let mut ctap = CtapState::new(&mut env, ...);  // first boot
ctap.foo(); // some commands (including configure)
let mut ctap = CtapState::new(&mut env, ...);  // second boot
ctap.bar(); // some other commands
```

Modifying directly the environment without going through the ctap library, should only be done for things that may be modified outside the library, e.g. customization (before any calls to the library), rng (the library can't guess the future), user presence (the library can't guess the user behavior), etc. The storage is not supposed to be modified outside the library.

Reply:
I think not changing `storage` under `CtapState`'s feet makes sense. The tests right now are a bit more concise than what you'd need to do ctap API only. Since we are building a helper function for that, it wouldn't be a readability advantage anymore, of course.

What we'd need to do is:
1. Prepare a ECDH handshake.
2. Use that to setup a PIN.
3. Use the PIN to get an AuthToken.
4. Call the custom `VendorConfig` command to set a certificate and attestation key.
5. Use the AuthToken to send an `AuthenticatorConfig` command.

Sounds like a lot, but steps 1-3 at least will be reusable for other tests. What we currently use is [mostly this](https://github.com/google/OpenSK/blob/a0e11bd5aa01bf73d9ccd4336524f96316962475/src/ctap/client_pin.rs#L559):

Reply:
Why do we need a PIN? The deploy script doesn't need it and simply calls the configure vendor command. Why wouldn't the following work?

```rust
ctap_state.process_vendor_configure(&mut env, params, channel)
```

But regardless of the answer, I think we agree that having a helper function for that is useful. The question is where should we put those helper functions. I guess a `#[cfg(feature = "std")]` module at the crate root: `test_helpers`, `test`, `test_helper`?

Reply:
1. `process_vendor_configure` is a private function. We have to call [process_parsed_command](https://github.com/google/OpenSK/blob/a0e11bd5aa01bf73d9ccd4336524f96316962475/src/ctap/mod.rs#L530). But that's a small difference.
2. Enabling attestation is not a feature of our own configuration command `VendorConfig`, but CTAP's [`AuthenticatorConfig`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#authenticatorConfig). I agree the naming is unfortunate. And this command must in almost all cases be protected by user verification.

And I agree it should be a `std` only module! I'd have liked to but it in [`tests`](https://doc.rust-lang.org/book/ch11-03-test-organization.html#the-tests-directory), but I guess the fuzzer can't access it there? So your ideas sound reasonable.

Reply:
I see. Then if I understand correctly, we would like the following (in possibly separate PRs or all in this PR):
- Create a `test` module at the root gated by the `std` feature.
- Add a `configure(&mut TestEnv, &mut CtapState, certificate, private_key)` command (bypassing HID and CBOR parsing, we might want wrapper functions for those alternatives) for `VendorConfig`.
- Add a `enable_enterprise_attestation(&mut TestEnv, &mut CtapState)` command for `AuthenticatorConfig`. This might need to deal with PIN and UV.
- Use those in tests and fuzzing.

From a distance, this looks like similar helpers one would need to write a blackbox CTAP test-suite like https://github.com/google/CTAP2-test-tool . Maybe we can have that in mind to help design.

Reply:
Resolved in doc.